File: 0x08-python-more_classes/4-rectangle.py
#!/usr/bin/python3
""" Rectangle class module. """


class Rectangle:
    """
    Initialises Rectangle attributes.
    """
    def __init__(self, width=0, height=0):
        self.width = width
        self.height = height

    """ Retrieves width property.
    """
    @property
    def width(self):
        return self.__width

    """ Retrieves height property.
    """
    @property
    def height(self):
        return self.__height

    """ Sets width of Rectangle.
    """
    @width.setter
    def width(self, value):
        if type(value) is not int:
            raise TypeError("width must be an integer")
        if value < 0:
            raise ValueError("width must be >= 0")

        self.__width = value

    """ Sets height of Rectangle.
    """
    @height.setter
    def height(self, value):
        if type(value) is not int:
            raise TypeError("height must be an integer")
        if value < 0:
            raise ValueError("height must be >= 0")

        self.__height = value

    """ Defines area of Rectangle.
    """
    def area(self):
        return self.__width * self.__height

    """ Defines perimeter of Rectangle.
    """
    def perimeter(self):
        if self.__width == 0 or self.__height == 0:
            return 0

        return 2 * (self.__width + self.__height)

    def __str__(self):
        if self.__width == 0 or self.__height == 0:
            return ''

        rect = ''
        for i in range(self.__height):
            for j in range(self.__width):
                rect = rect + '#'
            if i < self.__height - 1:
                rect = rect + '\n'

        return rect

    def __repr__(self):
        return "Rectangle({}, {})".format(self.__width, self.__height)

File: 0x13-javascript_objects_scopes_closures/4-rectangle.js
#!/usr/bin/node

class Rectangle {
  width;
  height;

  constructor (w, h) {
    if (w <= 0 || h <= 0) {
      return {};
    }
    if (typeof w === 'undefined' || typeof h === 'undefined') {
      return {};
    }

    this.width = w;
    this.height = h;
  }

  print () {
    for (let i = 0; i < this.height; i++) {
      let str = '';
      for (let j = 0; j < this.width; j++) {
        str += 'X';
      }
      console.log(str);
    }
  }

  rotate () {
    [this.height, this.width] = [this.width, this.height];
  }

  double () {
    this.height = this.height * 2;
    this.width = this.width * 2;
  }
}

module.exports = Rectangle;
